synapse/storage/databases/main/events_bg_updates.py

```diff
@@ -1127,48 +1137,71 @@ def _event_thread_relation_txn(txn: LoggingTransaction) -> int:
                     )
                     continue
 
-                # If there's no relation (or it is not a thread), skip!
+                # If there's no relation, skip!
                 relates_to = event_json["content"].get("m.relates_to")
                 if not relates_to or not isinstance(relates_to, dict):
                     continue
-                if relates_to.get("rel_type") != RelationTypes.THREAD:
+
+                # If the relation type or parent event ID is not a string a
+                # string, skip it.
+                #
+                # Do not consider relation types that have existed for a long time,
+                # only include the new thread relation and any unknown relations.
+                rel_type = relates_to.get("rel_type")
+                if not isinstance(rel_type, str) or rel_type in (
+                    RelationTypes.ANNOTATION,
+                    RelationTypes.REFERENCE,
+                    RelationTypes.REPLACE,
+                ):
                     continue
 
-                # Get the parent ID.
                 parent_id = relates_to.get("event_id")
                 if not isinstance(parent_id, str):
                     continue
 
-                missing_thread_relations.append((event_id, parent_id))
+                missing_relations.append((event_id, parent_id, rel_type))
+
+            # Insert the missing data, note that we upsert here in-case the event
+            # has already been processed.
+            if missing_relations:
+                self.db_pool.simple_upsert_many_txn(
+                    txn=txn,
+                    table="event_relations",
+                    key_names=("event_id",),
+                    key_values=[(r[0],) for r in missing_relations],
+                    value_names=("relates_to_id", "relation_type"),
+                    value_values=[r[1:] for r in missing_relations],
+                )
 
-            # Insert the missing data.
-            self.db_pool.simple_insert_many_txn(
-                txn=txn,
-                table="event_relations",
-                values=[
-                    {
-                        "event_id": event_id,
-                        "relates_to_Id": parent_id,
-                        "relation_type": RelationTypes.THREAD,
-                    }
-                    for event_id, parent_id in missing_thread_relations
-                ],
-            )
+                # Iterate the parent IDs and invalidate caches.
```

Comment:
I don't know if we usually clear caches as part of background updates, but in this case I think it could be a bit broken without doing it.

Reply:
seems sensible